rest/api.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.  // You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package rest
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus-community/promql-langserver/internal/vendored/go-tools/lsp/protocol"
+	"github.com/prometheus-community/promql-langserver/langserver"
+	promClient "github.com/prometheus-community/promql-langserver/prometheus"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/route"
+)
+
+func returnJSON(w http.ResponseWriter, content interface{}) {
+	encoder := json.NewEncoder(w)
+
+	err := encoder.Encode(content)
+	if err != nil {
+		http.Error(w, errors.Wrapf(err, "failed to write response").Error(), 500)
+	}
+}
+
+type lspData struct {
+	// Expr is the promQL expression and is required for all endpoint available.
+	Expr string `json:"expr"`
+	// Limit is the number max of result returned to the client. It will be used for the autocompletion and the diagnostics.
+	Limit *uint64 `json:"limit,omitempty"`
+	// PositionLine is the number of the line for which the metadata is queried.
+	PositionLine *float64 `json:"positionLine,omitempty"`
+	// PositionChar for which the metadata is queried. Characters are counted as UTF16 Codepoints.
+	PositionChar *float64 `json:"positionChar,omitempty"`
+}
+
+func (d *lspData) UnmarshalJSON(data []byte) error {
+	var tmp lspData
```

Comment:
Not sure `tmp` is a descriptive name for this. How about just `data` or so?

Reply:
Actually, do you even need `tmp`? I'd just directly try unmarshaling into `d` (cast to `plain`).

Reply:
yeah you are right. I don't need it